[doc] Clarify our system and software requirements

imphil opened the pull request:
Before, we didn't explicitly list our system requirements in a 
distribution-agnostic way. Instead, we provided a list of packages and a 
reference distribution (Ubuntu 16.04). Users of other distributions then 
had to either try, or manully cross-check if their setup works as expected.

With this change the documentation describes better what packages are 
required, and which version is required. This commit is meant to be a 
clarification without changing the status quo.

Notes:
* The list of package dependencies and their versions isn't bullet-proof,
  as packages can change behavior even in-between minor
  versions (if they use semantic versioning at all). The only way to be
  sure that a distribution works is testing it. We currently do all our
  tests on Ubuntu 16.04, hence we call this the reference platform.
* OpenTitan has generally been doing a very good job at keeping
  compatibility with as many distributions and setups as
  possible, and we have been very open to feedback to improve the
  development experience for our users. We will continue to do so, but
  need user input for this: we cannot provide automated testing for all
  possible setups.

imphil commented:
@martin-lueker @rasmus-madsen This PR is in preparation for the Ubuntu 18.04 discussion; it only clarifies the status quo. I'd still appreciate if you can have a look and approve this PR.

rswarbrick commented on doc/ug/install_instructions/index.md:
```diff
@@ -4,30 +4,83 @@ title: Install Build Requirements
-TODO: Be more specific about the system requirements, especially the Linux distribution.
+* Physical access to that machine, root permissions and a graphical environment are assumed in this guide.
+  However, with an appropriate setup, none of these things are strictly needed.
+* 60 GB or more of disk space, depending on the used EDA tools.
```
Grammar nit: This should be "on the EDA tools used".

rswarbrick commented on doc/ug/install_instructions/index.md:
```diff
@@ -4,30 +4,83 @@ title: Install Build Requirements
+    However, we cannot guarantee they can be used "out of the box" and might require updates of packages.
+    Please file a [GitHub issue](https://github.com/lowRISC/opentitan/issues) if you need help or would like to propose a change to increase compatibility with other distributions.
+    * RedHat Enterprise Linux (RHEL)/CentOS 7 and 8
+    * Many of our contributors use rolling releaese distributions, which generally include the latest stable releases of all software packages, such as Debian unstable, openSUSE Tumbleweed, or Arch Linux.
```
Nit: Typo in "release"

rswarbrick commented on doc/ug/install_instructions/index.md:
```diff
@@ -4,30 +4,83 @@ title: Install Build Requirements
+* OpenSSL 1.0 or 1.1 with development headers
+* libusb 1.0
+* OpenOCD
+* [Verible](https://github.com/google/verible) {{< tool_version "verible" >}}
```
I like the `tool_version` shortcode! Would it make sense to use it for `ninja` too?

rswarbrick reviewed (approved):
This looks good. Would it make sense to leave a note in our apt-requirements file to update the install instructions if we add a new dependency (or changed version)?

imphil commented:
Thanks for having a look @rswarbrick. I updated the documentation with your comments, and also improved the wording around RHEL/CentOS.

lenary commented on apt-requirements.txt:
```diff
@@ -3,6 +3,8 @@
 
 # Keep sorted.
+# When updating this file also check if doc/ug/install_instructions/index.md
+# needs to be updated as well.
```
Is it worth pointing out that this list does get included in the install instructions where apt is mentioned?

lenary commented on doc/ug/install_instructions/index.md:
```diff
@@ -4,30 +4,89 @@ title: Install Build Requirements
+  Additional Python dependencies are installed through pip.
+* A C++14 capable compiler.
+  GCC 5 or Clang 3.5 should meet this requirement.
+* clang-format
```
Somewhere in the docs, we say exactly what version is required. Right now, I believe it's `3.8.x`. 

I think the "somewhere" is in the `utils/container/README.md`, which might need merging with this?

imphil replied:
We don't require a specific version at the moment, whatever is available as part of a supported LLVM version is fine. So effectively clang-format 3.5+ at the moment.

lenary replied:
Not true. More recent clang-format versions have subtle differences to the results of running clang-format in CI. See for instance #2035 and #2082. This is an issue googlers have run into, which is why a specific version of clang-format was documented at all.

imphil replied:
Can you point me to this documentation?

Yes, we use a specific version in CI and using another version will be painful on occasion, but I don't think we have a policy requiring a specific version at this point.

imphil replied:
I added a sentence on clang-format 3.8 being recommended, please let me know if that addresses your concerns @lenary.

lenary reviewed (commented):
Some nits. Let's not duplicate this info between the documentation for the utilities container and the install instructions, which have quite a bit of overlap.

imphil commented:
> Let's not duplicate this info between the documentation for the utilities container and the install instructions, which have quite a bit of overlap.

I updated the container README and also fixed some inconsistencies along the way, PTAL.

lenary reviewed (approved):
LGTM modulo documenting the clang-format version.

rasmus-madsen reviewed (approved):
LGTM
